cms/djangoapps/contentstore/tests/test_course_settings.py

```diff
@@ -971,25 +973,35 @@ def test_advanced_components_munge_tabs(self):
         course = modulestore().get_course(self.course.id)
         self.assertNotIn(open_ended_tab, course.tabs)
         self.assertNotIn(peer_grading_tab, course.tabs)
-        self.assertNotIn(notes_tab, course.tabs)
-
-    def mark_wiki_as_hidden(self, tabs):
-        """ Mark the wiki tab as hidden. """
-        for tab in tabs:
-            if tab.type == 'wiki':
-                tab['is_hidden'] = True
-        return tabs
-
-    def test_advanced_components_munge_tabs_hidden_tabs(self):
-        updated_tabs = self.mark_wiki_as_hidden(self.course.tabs)
-        self.course.tabs = updated_tabs
+        self.assertNotIn(self.notes_tab, course.tabs)
+
+    def test_advanced_components_munge_tabs_validation_failure(self):
+        with patch('contentstore.views.course._refresh_course_tabs', side_effect=InvalidTabsException):
+            resp = self.client.ajax_post(self.course_setting_url, {
+                ADVANCED_COMPONENT_POLICY_KEY: {"value": ["notes"]}
+            })
+            self.assertEqual(resp.status_code, 400)
+
+            error_msg = [{'message': '', 'model': {'display_name': 'Tabs Exception'}}]
+            self.assertEqual(json.loads(resp.content), error_msg)
+
+            # verify that the course wasn't saved into the modulestore
+            course = modulestore().get_course(self.course.id)
+            self.assertNotIn("notes", course.advanced_modules)
+
+    @ddt.data(
```

Comment:
Sorry, that wasn't clear. This test is great, and I think it covers the cases we're interested in. Maybe it should also verify that only the one new tab is added, because before these changes the code would have added in all sorts of new tabs (wiki, discussion etc).

Thanks for the pointer to the other tests. This is what I had in mind, but I think we should extend ``invalid_tabs`` to cover more of the cases we've run into:

https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/tests/test_tabs.py#L476

So, add duplicate tabs, unknown tabs, wrong tab order etc.